h2/connection.py

```diff
@@ -1634,6 +1634,33 @@ def _receive_push_promise_frame(self, frame):
 
         return frames, events + stream_events
 
+    def _handle_data_on_closed_stream(self, events, exc, frame):
+        # This stream is already closed - and yet we received a DATA frame.
+        # The received DATA frame counts towards the connection flow window.
+        # We need to manually to acknowledge the DATA frame to update the flow
+        # window of the connection. Otherwise the whole connection stalls due
+        # the inbound flow window being 0.
+        if self._stream_is_closed_by_reset(exc.stream_id):
```

Comment:
I don't think that this condition is useful. When we receive `DATA` frame for a closed stream, this is just a stream error, connection is still valid, and I think that we have to acknowledge received data without any conditions.

There is a possibility that we don't know why stream was closed – `Connection._closed_streams` has a limit, so I propose to simplify the logic and remove this condition.

Reply:
This would mean a changing behaviour that is currently ensured by this test:
https://github.com/python-hyper/hyper-h2/blob/aa66d45c177e585259834eae61c2fd90e0f760a5/test/test_closed_streams.py#L236

If I remove the condition, we will never raise anything else - and therefore not product a ProtocolError...

Reply:
> If I remove the condition, we will never raise anything else - and therefore not product a ProtocolError...

There are two types of errors in the spec:
- [5.4.1. Connection Error Handling](https://httpwg.org/specs/rfc7540.html#rfc.section.5.4.1)
- [5.4.2. Stream Error Handling](https://httpwg.org/specs/rfc7540.html#rfc.section.5.4.2)

Connection error means that connection is broken and should be closed. Stream error means that connection still can be used.

I believe that `Connection.receive_data()` should raise `ProtocolError` only for connection errors.

Here are another interesting references to the spec:

- https://httpwg.org/specs/rfc7540.html#rfc.section.6.1
   > DATA frames are subject to flow control and can only be sent when a stream is in the "open" or "half-closed (remote)" state. The entire DATA frame payload is included in flow control, including the Pad Length and Padding fields if present. If a DATA frame is received whose stream is not in "open" or "half-closed (local)" state, the recipient MUST respond with a **stream error (Section 5.4.2) of type STREAM_CLOSED**.

- https://httpwg.org/specs/rfc7540.html#rfc.section.6.9
   > A receiver that receives a flow-controlled frame MUST always account for its contribution against the connection flow-control window, unless the receiver treats this as a connection error (Section 5.4.1). **This is necessary even if the frame is in error.** The sender counts the frame toward the flow-control window, but if the receiver does not, the flow-control window at the sender and receiver can become different.

So maybe those tests don't take this into account.